core/async/src/examples/simple.rs

```diff
@@ -0,0 +1,64 @@
+use std::{sync::Arc, time::Duration};
+
+use derive_enum_from_into::{EnumFrom, EnumTryInto};
+
+use crate::{
+    messaging::Sender,
+    test_loop::{CaptureEvents, LoopEventHandler, TestLoopBuilder, TryIntoOrSelf},
+};
+
+use super::sum_numbers::{ReportSumMsg, SumNumbersComponent, SumRequest};
+
+#[derive(derive_more::AsMut, derive_more::AsRef)]
+struct TestData {
+    summer: SumNumbersComponent,
+    sums: Vec<ReportSumMsg>,
+}
+
+#[derive(Debug, EnumTryInto, EnumFrom)]
+enum TestEvent {
+    Request(SumRequest),
+    Sum(ReportSumMsg),
+}
+
+// Handler that forwards SumRequest messages to the SumNumberComponent.
+// Note that typically we would have a single handler like this, and it can
+// be reused for any test that needs to send messages to this component.
+pub struct ForwardSumRequest;
+
+impl<Data: AsMut<SumNumbersComponent>, Event: TryIntoOrSelf<SumRequest>>
+    LoopEventHandler<Data, Event> for ForwardSumRequest
+{
+    fn handle(&mut self, event: Event, data: &mut Data) -> Option<Event> {
+        match event.try_into_or_self() {
+            Ok(request) => {
+                data.as_mut().handle(request);
+                None
+            }
+            Err(event) => Some(event),
+        }
+    }
+}
+
+#[test]
+fn test_simple() {
+    let builder = TestLoopBuilder::<TestEvent>::new();
+    // Build the SumNumberComponents so that it sends messages back to the test loop.
+    let data =
+        TestData { summer: SumNumbersComponent::new(Arc::new(builder.sender())), sums: vec![] };
+    let sender = builder.sender();
+    let mut test = builder.build(data);
+    test.register_handler(ForwardSumRequest);
+    test.register_handler(CaptureEvents::<ReportSumMsg>::new());
+
+    sender.send(TestEvent::Request(SumRequest::Number(1)));
+    sender.send(TestEvent::Request(SumRequest::Number(2)));
+    sender.send(TestEvent::Request(SumRequest::GetSum));
+    sender.send(TestEvent::Request(SumRequest::Number(3)));
+    sender.send(TestEvent::Request(SumRequest::Number(4)));
+    sender.send(TestEvent::Request(SumRequest::Number(5)));
+    sender.send(TestEvent::Request(SumRequest::GetSum));
+
+    test.run(Duration::from_millis(1));
```

Comment:
uh.. I really don't like when we pass 'time' into the tests - as this can make them flaky.

Could we have a different solution?

Reply:
Just wanna make sure our understanding is on the same page first. The test loop maintains a virtual time, and each event has a due time. If we don't use any timers at all, or have any simulated network delays, then all events have due time 0. When we pass a deadline of 1ms, it simply means execute all events that have due time 0. If we do have timers or simulated network delay, then the deadline controls when we stop simulating, based on the virtual time that we calculate deterministically, and we'll never depend on wall time for this.

For example, let's say we test a client actor and the client actor has a timer that fires every 100ms checking for certain conditions. The way we'd implement that timer in this framework is to have an event that is initially enqueued with due time = 100ms, and whenever it is handled (which is guaranteed to be after every other event with <100ms due time), it enqueues another event with 100ms delay (i.e. next event will (precisely) have a 200ms due time, next 300ms, etc.). But this never ends because it's a repeating timer. So to avoid going into an infinite loop, my solution here is to specify a deadline, and the intent is that the deadline should be specified to be pretty long so that we are obviously going to make it.

Timers are important because certain actions, like refetching chunks in ShardsManager, are required to complete a task. There are a couple other ideas for managing this:
 1. Manually trigger timers by sending an event like "FireAllTimers" - problem here is, the primary motivation for this framework is to avoid having to manually trigger events, which is hard to do right. The fact that we need a timer to complete something is a bit of an implementation detail.
 2. Trigger timers automatically, but if we reach a point where the sequence of events executed ends with a contiguous sequence of only timer events and these events cover all timers registered on the event loop, then halt. In other words, if running timers no longer yield anything other than timers, then we're kinda done. The problem is that this is only an approximation, since some timers fire more frequently than they might actually do something (like firing every 100ms just to check if some chunk is due to be requested again).

I'm open to other suggestions here.